fms/modules/attention.py

```diff
@@ -168,13 +170,24 @@ def forward(
                 )
 
         # if you want to use caching and past_key_value_state is not None meaning you have values in your cache
-        if use_cache and past_key_value_state is not None:
-            if is_self:
-                keys = torch.cat((past_key_value_state[0], keys), dim=2)
-                values = torch.cat((past_key_value_state[1], values), dim=2)
+        if use_cache:
+            # we already have a cache so add to it or get it
+            if past_key_value_state is not None:
+                if is_self:
+                    keys = torch.cat((past_key_value_state[0], keys), dim=2)
+                    values = torch.cat((past_key_value_state[1], values), dim=2)
+                else:
+                    keys = past_key_value_state[0]
+                    values = past_key_value_state[1]
+            # we do not have a cache so must initialize it
             else:
-                keys = past_key_value_state[0]
-                values = past_key_value_state[1]
+                preallocate_length = int(math.pow(2, int(math.log2(keys.size(2))) + 1))
+                keys = ExpandableTensor(
+                    keys, dim=2, preallocate_length=preallocate_length
+                )
+                values = ExpandableTensor(
```

Comment:
this effectively makes them non-contiguous. I wonder if anywhere thye're being used will take a copy when using them, i.e. if it would be better to wrap them somewhere further down.

Reply:
as discussed, this will only be on the first iteration of a use_cache usage. others will be expandabletensor anyway so not worth trying to change this wrap.